Update Working/Project Group RFC (#32)
rust-lang/wg-governance: draft-rfcs/working-group-terminology.md

# Summary

> Currently the Rust Programming Language organisation has a set of teams

Reviewer: I tend to favor "bullet point" summaries that try to capture the important changes being proposed (and in particular that do not contain much motivation). I took a stab in [this hackmd](https://hackmd.io/xfW6jj7DQXeeVeNoCW2g3g) at defining some of the key points from this RFC, inlined below. Maybe something like this?

* Formalize **project groups** as groups dedicated to completing specific projects within the context of a Rust team
    * Project groups are created via an RFC and have a "parent team" (or multiple teams)
    * The groups then drive the project to completion, e.g. by authoring follow-up RFCs and doing design work
    * Each project group typically has:
        * charter
        * defined shepherds and team liasons
        * an associated repository
        * regular meetings
        * dedicated streams on Discord/Zulip/etc
* Define **working groups** to refer to the "domain working groups" that are created to explore particular domains, such as embedded, CLI, etc
    * They have a charter and defined leads but operate more independently from the Rust teams
* Define **community group** as the term for groups not formally affiliated with the Rust project

> Currently there the Rust Programming Language organisation has a set of teams

Reviewer 2: Looks like a previous PR accidentally introduced a superfluous word.

```suggestion
Currently the Rust Programming Language organisation has a set of teams
```

## Common Aspects of Working Groups and Project Groups

Before going into the differences between these groups it's important to start
with what it is shared between them.

- A charter that defines the scope and intent of the group.

- A GitHub repository hosted under the `rust-lang` organization
  (or separate organisation if necessary) containing the charter and
  instructions for how community members can monitor or participate in
  the group.

- Groups have at least one shepherd who acts as the leader of the group and is
  typically responsible for writing the initial charter, handling administrative
  and communication tasks, as well as delegating responsibilities to other
  members in the group.

- Groups should have a liaison member associated with an official Rust team.

  - This liaison may also be but is not required to be a shepherd.

- Groups and their members are represented on the official
  [rust-lang.org] website.

- Membership requirements for both groups is decided by the shepherd and should
  be stated in the charter.

  - Initial membership should try to represent people who have already been
    participating regularly and productively in the respective area.

> - Neither group has _"formal decision making power"_. Where "formal decision

Reviewer 2: ```suggestion
- Neither group has _"formal decision making power"_: this means that they cannot unilaterally accept (i.e. merge) RFCs on...
```

  making power" is defined as being able to accept RFCs on `rust-lang/rfcs`.
  Similarly, neither group has representation on the Core team.

>   - Groups are of course however able to and encouraged to create RFCs as well

Reviewer 2: ```suggestion
  - Groups are of course encouraged to create RFCs as well
```

    as advocate their concerns and desired changes to the Rust teams
    and community.

- Both groups are entitled to spaces in any or all of Rust's officially managed
  discussion forums.

  - As the time of this writing this includes [Zulip] and [Discord].
  - Groups are also free to create spaces on platforms where the Rust project
    does not have a official presence. Moderation for these spaces should still
    follow Rust's [Code of Conduct].

[rust-lang.org]: https://rust-lang.org
[zulip]: https://rust-lang.zulipchat.com
[discord]: https://discord.gg/rust-lang
[code of conduct]: https://www.rust-lang.org/policies/code-of-conduct

### Initial Setup

Once a group has been approved, a pull request with the initial set of members
should be made to [`rust-lang/team`]. Please refer to team's documentation for
how to create a group.

It is recommended for the working group to create a `wg-{name}`/`project-{name}`
repository under the `rust-lang` organisation using the [project group
template], and making any relevant changes and personalisation.

[`rust-lang/team`]: https://github.com/rust-lang/team

> [project group template]: https://github.com/rust-lang/project-group-template

Reviewer 2: Will working groups and project groups share a template? (I don't see a specific reason why not, but I want to make sure that's the intention.)

Author: Yeah that is my intention currently.

[Game Development][gamedev-wg].

[embedded-wg]: https://github.com/rust-embedded/wg
[wasm-wg]: https://github.com/rustwasm/team
[gamedev-wg]: https://github.com/rust-gamedev/wg

### Working Group Creation

Creation of a working group is approved by the core team. Typically this has
been done by the core team agreeing to approve the creation of new working
groups and having a period of time soliciting applications from the community,
and then approving a subset of those applications.

> **DRAFT NOTE** Should this application come in the form of an RFC? My
> inclination is yes, however it could just create more churn and drama than
> needed. Posting in a thread on internals as was done previously might
> be enough.

#### Application Checklist

This not meant to be formal list of questions to be answer, however the
application should try to address these points.

- What value do you see your group bringing to the organisation?
- Why should the project your group?
- What are the goals of your group?
- What are explicitly non-goals of your group?

These first four points should also serve as the basis for the charter for the
working group.

- Is your group long-running or temporary?

>   - If it is temporary, how long do you see it running for?

Reviewer 2: If the answer is "temporary", I suspect that might be a sign that the group would be better organized as a project group. Is that worth mentioning here? (I don't mean to imply that working groups should never be temporary, just that that's probably rarer than temporary project groups.)

Author: Well no, because a project group would be part of a team, and are in a way managed by the team, where as a working group is usually self managed.

- What is the long-term vision of your group?

> - What do you expect the relationship to the organisation be?

Reviewer 3: Folks can have a hard time coming to a shared understanding of terms like vision and mission, and it feels like there is a big gap between "short-term goals" and "long-term vision". Describing group activities in a more explicitly structured way could be helpful.  It might also be useful to put the current contents of lines 161 and 166 next to each other to help folks integrate the different levels of their statement of purpose. Also, examples may be helpful. New version of lines 161 and 166 with an additional middle level below:

What vision does your group want to realize?
Example: Official Rust groups are organized in a sustainable, clearly described way.

What are some of the core objectives that support that vision? 
Example: Creating a clear framework for group types and ensuring that group-related infrastructure is up to date. 

What are your group's short-term goals (for its first month or first year, depending on group scope)? Creating an RFC describing group types and the differences between them; Auditing existing repositories for recent activity.

These examples aren't perfect, but hopefully they provide some guidance. They can of course be tweaked or replaced.

- How do you want to establish accountability?
- If applicable, which other working groups or teams do you expect to have close
  contact with?
- What are the short-term goals of your group?
- Who are the initial shepherds/leaders? (This is preferably 2–3 individuals,
  but not required.)
- How do you intend to make your work accessible to people outside your group?
- Where do you see your group needing help?
- What is your preferred way for the core team to contact you?

### Working Group Evaluation

Working group shepherds should be regularly participating in the project leads
discussion channels so that the rest of the organisation is aware of the work
and can help where needed.

If a working group has demonstrated consistent productivity over a significant
period time, and there is consensus that there is significant future work, it
may become a Rust team. Conversely if there is consensus that the work is
"complete" to the point that there's there is little benefit to continuing the

> working group. It may be wound down.

Reviewer 2: ```suggestion
working group, it may be wound down.
```

The wind down process of a working group involves communicating the wind down to
the community and the archival or transfer of ownership of the relevant projects.

> ## Project Groups

Reviewer 2: Do we want to formalize the notion that Project groups are intended to be the "canonical" way for "pre-RFC" collaboration to occur?

Author: Well I think we need more time with it. We can say this is some teams intend to have pre-RFC discussions for now, but I would want to have a couple of retrospectives and see if there are issues when others try to adopt it before we give a "canonical" endorsement.

A Project Group is a group of people working on a particular project or
responsibilities at the behest of an official Rust team. Examples of this would

> include [FFI Unwind], [Inline ASM], or [Safe Transmute].

Reviewer 2: ```suggestion
include [FFI Unwind], [Inline ASM], and [Safe Transmute].
```

> The goal of a project is build a community or formalise and existing community

Reviewer 3: A community or a team?

> The goal of a project is build a community or formalise and existing community

Reviewer: To me, the primary goal of a project group is to ensure **follow through**. That is, it has the responsibility for designing a feature and then seeing that through all the way to stabilization. It is also a way to build a community around the feature, which can be helpful both for sharing the work but also for doing follow-up work related to that feature.

around a particular feature or project in the in the organisation, and use this
space to discuss and iterate on that feature.

> Part of building a community is removing some of the institutional memory that

Reviewer 3: Not removing, perhaps rather stewarding? Providing a home for this discussion rather than having it end up buried in the RFC thread?

develops in the design process, and centralising the information and discussion
around the feature so that we can provide better visibility into why certain
decisions and trade offs were made over others.

Previously a lot of the discussion and iteration for large features would
happen in the initial RFC thread. This leads to a lot of discussion in the top
of the thread and that has can become completely irrelevant to the current
iteration.

This process has also been unsuitable to describe features that can take
multiple years to develop and will become multiple RFCs over the course of its
design process. Some examples of of this are the "`impl Trait`" and "macros
2.0" features, where the goals has shifted a lot from the initial RFCs, and it
can be hard to know their current status.

### Project Group Creation

Since project groups are approved by their relevant parent team over the core
team, it's up to each team decide their specific requirements. However we
recommend using the [application checklist](#application-checklist) as the basis

> for process and if needed adding any extra requirements.

Reviewer: One thing we don't mention here is project group *membership*, right?

I think we want to say that project groups have a notion of members that is maintained at the shepherds' discretion. Typically, people who are productively contributing to the project group for some time will be added as members. But it is not required or expected that all project groups will be large: smaller features may only need one or two people, including the shepherd plus team liasons.

### Project Group Evaluation

Parent teams should add checking in with their project groups as part of their
regular triage.

[ffi unwind]: https://github.com/rust-lang/project-ffi-unwind
[inline asm]: https://github.com/rust-lang/project-inline-asm
[safe transmute]: https://github.com/rust-lang/project-safe-transmute

> 

Reviewer 3: What is the process for closing/winding down a project group? Is this vision realized, objectives completed, something else? At what level does a group's core and unchanging purpose reside--vision? How to ensure that this vision is realizable? There will be much inertia around declaring a project group's work done and actually winding down the group.

## Community Groups

Community Groups are groups of individuals who want to work and collaborate on a
specific topic, without going through the formal process of a working group.
Community groups are encouraged to adopt the structure of working or projects
groups laid, but are free create and experiment with their own structure. As
such community groups are not officially endorsed by The Rust Programming
Language Organisation.

## Retrospectives

While this RFC attempts to address some of the current organisational problems
within the organisation, it also doesn't believe that this RFC will be a panacea
to those problems or that we won't encounter more in the future. As part of
that, we'd also like to introduce performing retrospectives with groups, once
significant time has past or the group has been finished it's project.

This would involve a discussion between the members of the group, their parent
team, and the Governance working group. The retrospective should produce a
public blog post on the Inside Rust blog, however any feedback a member has that

> they would want to keep private would be omitted.

Reviewer 3: More structure for the retrospective might be useful.
